adapters/rubicon/rubicon.go

```diff
@@ -941,6 +909,104 @@ func resolveBidFloor(bidFloor float64, bidFloorCur string, reqInfo *adapters.Ext
 	return bidFloor, nil
 }
 
+func updateImpRpTargetWithFpdAttributes(extImp openrtb_ext.ExtImpRubicon, imp openrtb2.Imp,
+	site *openrtb2.Site, app *openrtb2.App) (json.RawMessage, error) {
+	existingTarget, _, _, _ := jsonparser.Get(imp.Ext, "rp", "target")
+
+	target := populateFirstPartyDataAttributes(extImp.Inventory, existingTarget)
+
+	if site != nil {
+		siteExtData, _, _, _ := jsonparser.Get(site.Ext, "data")
+		target = populateFirstPartyDataAttributes(siteExtData, target)
+		if len(site.SectionCat) > 0 {
+			target = addStringArrayAttribute(site.SectionCat, target, "sectioncat")
+		}
+		if len(site.PageCat) > 0 {
+			target = addStringArrayAttribute(site.PageCat, target, "pagecat")
+		}
+		if site.Page != "" {
+			target = addStringAttribute(site.Page, target, "page")
+		}
+		if site.Ref != "" {
+			target = addStringAttribute(site.Ref, target, "ref")
+		}
+		if site.Search != "" {
+			target = addStringAttribute(site.Search, target, "search")
+		}
+	} else {
+		appExtData, _, _, _ := jsonparser.Get(app.Ext, "data")
+		target = populateFirstPartyDataAttributes(appExtData, target)
+		if len(app.SectionCat) > 0 {
+			target = addStringArrayAttribute(app.SectionCat, target, "sectioncat")
+		}
+		if len(app.PageCat) > 0 {
+			target = addStringArrayAttribute(app.PageCat, target, "pagecat")
+		}
+	}
+
+	impExtContextAttributes, _, _, _ := jsonparser.Get(imp.Ext, "context", "data")
+	impExtDataAttributes, _, _, _ := jsonparser.Get(imp.Ext, "data")
+
+	if impExtContextAttributes != nil && len(impExtContextAttributes) > 0 {
+		target = populateFirstPartyDataAttributes(impExtContextAttributes, target)
+	} else if impExtDataAttributes != nil && len(impExtDataAttributes) > 0 {
+		target = populateFirstPartyDataAttributes(impExtDataAttributes, target)
+	}
+
+	if len(extImp.Keywords) > 0 {
+		target = addStringArrayAttribute(extImp.Keywords, target, "keywords")
+	}
+
+	return target, nil
+}
+
+func addStringAttribute(attribute string, target json.RawMessage, attributeName string) json.RawMessage {
+	targetAsMap := messageToMap(target)
+	targetAsMap[attributeName] = [1]string{attribute}
+
+	updatedTarget, _ := json.Marshal(targetAsMap)
+	return updatedTarget
+}
+
+func addStringArrayAttribute(attribute []string, target json.RawMessage, attributeName string) json.RawMessage {
+	targetAsMap := messageToMap(target)
+	targetAsMap[attributeName] = attribute
+
+	updatedTarget, _ := json.Marshal(targetAsMap)
+	return updatedTarget
+}
+
+func getAdSlot(imp openrtb2.Imp) string {
+	adServerContextName, _ := jsonparser.GetString(imp.Ext, "context", "data", "adserver", "name")
+
+	if adServerContextName == "gam" {
+		contextAdSlot, _ := jsonparser.GetString(imp.Ext, "context", "data", "adserver", "adslot")
+		return contextAdSlot
+	}
+	adServerDataName, _ := jsonparser.GetString(imp.Ext, "data", "adserver", "name")
+	if adServerDataName == "gam" {
+		dataAdSlot, _ := jsonparser.GetString(imp.Ext, "data", "adserver", "adslot")
```

Comment:
Please do not ignore errors and add appropriate error handling logic here

Reply:
Same as above